compiler/rustc_hir_analysis/src/check/always_applicable.rs

```diff
@@ -0,0 +1,195 @@
+use rustc_data_structures::fx::FxHashSet;
+use rustc_errors::{struct_span_err, ErrorGuaranteed};
+use rustc_infer::infer::outlives::env::OutlivesEnvironment;
+use rustc_infer::infer::{RegionResolutionError, TyCtxtInferExt};
+use rustc_middle::ty::subst::SubstsRef;
+use rustc_middle::ty::util::CheckRegions;
+use rustc_middle::ty::{self, Ty, TyCtxt};
+use rustc_trait_selection::traits::{self, ObligationCtxt};
+
+use crate::hir::def_id::DefId;
+
+/// This function confirms that the trait implementation identified by
+/// `trait_impl_def_id` is not any more specialized than the type it is
+/// attached to (Issue #8142).
+///
+/// This means:
+///
+/// 1. The generic region/type parameters of the impl's self type must
+///    all be parameters of the Trait impl itself (i.e., no
+///    specialization like `impl Trait for Foo<i32>`), and,
+///
+/// 2. Any bounds on the generic parameters must be reflected in the
+///    struct/enum definition for the nominal type itself (i.e.
+///    cannot do `struct S<T>; impl<T:Clone> Trait for S<T> { ... }`).
+///
+pub fn check_trait_impl(tcx: TyCtxt<'_>, trait_impl_def_id: DefId) -> Result<(), ErrorGuaranteed> {
+    let self_type = tcx.type_of(trait_impl_def_id).subst_identity();
+    check_trait_impl_given_ty(tcx, trait_impl_def_id, self_type)
+}
+
+fn check_trait_impl_given_ty<'tcx>(
+    tcx: TyCtxt<'tcx>,
+    trait_impl_def_id: DefId,
+    ty: Ty<'tcx>,
+) -> Result<(), ErrorGuaranteed> {
+    match ty.kind() {
+        ty::Adt(adt_def, adt_to_impl_substs) => {
+            ensure_trait_params_and_item_params_correspond(
+                tcx,
+                trait_impl_def_id,
+                adt_def.did(),
+                adt_to_impl_substs,
+            )?;
+
+            ensure_trait_predicates_are_implied_by_item_defn(
+                tcx,
+                trait_impl_def_id,
+                adt_def.did(),
+                adt_to_impl_substs,
+            )
+        }
+
+        ty::Ref(_, ty, _) | ty::Array(ty, _) | ty::Slice(ty) => {
```

Comment:
Hm, @nikomatsakis may have thought more about this so idk if y'all have discussed these rules... but I'm not exactly sure what the "always applicable" rules should be for fundamental types like ref. I would've thought that refs kinda act like an ADT defined like:

`struct Ref<'a, T>(&'a T);`

So that the only valid impls for those would be those that look like `impl<'a, T: ?Sized> !Trait for &'a T`?

Reply:
Also, why is the region in `ty::Ref` and the array length in `ty::Array` being thrown away? Those both affect the applicability of the negative impls here.